File: in_Java/0155 Min Stack.java
class MinStack {

    Stack<Integer> st;
    int min;
    /** initialize your data structure here. */
    public MinStack() {
        st = new Stack<>();
        min = Integer.MAX_VALUE;
    }
    
    public void push(int x) {
        if(x <= min){
            st.push(min);
            min = x;
        }
        st.push(x);
    }
    
    public void pop() {
        if(st.peek() == min){
            st.pop();
            min = st.pop();
        }else{
            st.pop();
        }
    }
    
    public int top() {
        return st.peek();
    }
    
    public int getMin() {
        return min;
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(x);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */

File: in_Python/0155 Min Stack.py
class MinStack(object):

    def __init__(self):
        """
        initialize your data structure here.
        """
        self.arr = []
        self.minv = sys.maxint 
        

    def push(self, x):
        """
        :type x: int
        :rtype: None
        """
        if x <= self.minv:
            self.arr.append(self.minv)
            self.minv = x
        self.arr.append(x)

    def pop(self):
        """
        :rtype: None
        """
        if self.arr.pop() == self.minv:
            self.minv = self.arr.pop()

    def top(self):
        """
        :rtype: int
        """
        return self.arr[-1]

    def getMin(self):
        """
        :rtype: int
        """
        return self.minv
        


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(x)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()